File: android/app/src/main/java/com/example/lottery/util/Match.kt
package com.example.lottery.util

import kotlin.math.min

/**
 * 中奖匹配与等级计算（从网页 utils/match.js 原样移植，规则与官方一致）。
 * 号码均为逗号分隔的字符串，例如 "03,08,15,20,21,24"。
 */
object Match {

    fun splitNums(s: String?): List<String> =
        (s ?: "").split(",").map { it.trim() }.filter { it.isNotEmpty() }

    fun intersectCount(a: List<String>, b: List<String>): Int {
        val set = b.toSet()
        return a.count { set.contains(it) }
    }

    data class HitBalls(val hitRed: List<String>, val hitBlue: List<String>)

    fun hitBalls(userRed: String?, userBlue: String?, drawRed: String?, drawBlue: String?): HitBalls {
        val dr = splitNums(drawRed).toSet()
        val db = splitNums(drawBlue).toSet()
        return HitBalls(
            splitNums(userRed).filter { dr.contains(it) },
            splitNums(userBlue).filter { db.contains(it) }
        )
    }

    data class TierResult(val mr: Int, val mb: Int, val tier: String, val won: Boolean)

    /**
     * 计算中奖等级。
     * poolAmount: 该期奖池金额（元），用于判断福运奖是否生效（SSQ奖池≥15亿才触发）。
     */
    fun matchTier(type: String, userRed: String?, userBlue: String?, drawRed: String?, drawBlue: String?, poolAmount: Long = 0): TierResult {
        val mr = intersectCount(splitNums(userRed), splitNums(drawRed))
        val mb = intersectCount(splitNums(userBlue), splitNums(drawBlue))
        var tier = ""
        if (type == "ssq") {
            val fyjActive = poolAmount >= 15_0000_0000 // 15亿元
            tier = when {
                mr == 6 && mb == 1 -> "一等奖"
                mr == 6 && mb == 0 -> "二等奖"
                mr == 5 && mb == 1 -> "三等奖"
                (mr == 5 && mb == 0) || (mr == 4 && mb == 1) -> "四等奖"
                (mr == 4 && mb == 0) || (mr == 3 && mb == 1) -> "五等奖"
                mb == 1 -> "六等奖"
                mr == 3 && mb == 0 && fyjActive -> "福运奖"
                else -> ""
            }
        } else {
            tier = when {
                mr == 5 && mb == 2 -> "一等奖"
                mr == 5 && mb == 1 -> "二等奖"
                (mr == 5 && mb == 0) || (mr == 4 && mb == 2) -> "三等奖"
                mr == 4 && mb == 1 -> "四等奖"
                (mr == 4 && mb == 0) || (mr == 3 && mb == 2) -> "五等奖"
                (mr == 3 && mb == 1) || (mr == 2 && mb == 2) -> "六等奖"
                (mr == 3 && mb == 0) || (mr == 2 && mb == 1) || (mr == 1 && mb == 2) || (mr == 0 && mb == 2) -> "七等奖"
                else -> ""
            }
        }
        return TierResult(mr, mb, tier, tier.isNotEmpty())
    }

    data class PlayConfig(
        val minRed: Int, val maxRed: Int, val minBlue: Int, val maxBlue: Int,
        val redCount: Int, val blueCount: Int, val name: String,
        val redLabel: String, val blueLabel: String
    )

    fun playConfig(type: String): PlayConfig = if (type == "dlt") {
        PlayConfig(5, 12, 2, 12, 35, 12, "大乐透", "前区", "后区")
    } else {
        PlayConfig(6, 20, 1, 16, 33, 16, "双色球", "红球", "蓝球")
    }

    private fun combinations(arr: List<String>, k: Int): List<List<String>> {
        val res = mutableListOf<List<String>>()
        val n = arr.size
        if (k < 0 || k > n) return res
        val idx = IntArray(k)
        fun rec(start: Int, i: Int) {
            if (i == k) {
                res.add(idx.map { arr[it] })
                return
            }
            var j = start
            while (j < n) {
                idx[i] = j
                rec(j + 1, i + 1)
                j++
            }
        }
        rec(0, 0)
        return res
    }

    private fun diff(a: List<String>, b: List<String>): List<String> {
        val set = b.toSet()
        return a.filter { !set.contains(it) }
    }

    fun enumerateCombos(type: String, red: String?, blue: String?, bankerRed: String?, bankerBlue: String?): List<Pair<List<String>, List<String>>> {
        val cfg = playConfig(type)
        val reds = splitNums(red)
        val blues = splitNums(blue)
        val bRed = splitNums(bankerRed)
        val bBlue = splitNums(bankerBlue)
        val redCombos = if (bRed.isNotEmpty()) {
            combinations(diff(reds, bRed), cfg.minRed - bRed.size).map { bRed + it }
        } else {
            combinations(reds, cfg.minRed)
        }
        val blueCombos = if (bBlue.isNotEmpty()) {
            combinations(diff(blues, bBlue), cfg.minBlue - bBlue.size).map { bBlue + it }
        } else {
            combinations(blues, cfg.minBlue)
        }
        val out = mutableListOf<Pair<List<String>, List<String>>>()
        for (rc in redCombos) for (bc in blueCombos) out.add(Pair(rc, bc))
        return out
    }

    fun ticketBets(type: String, red: String?, blue: String?, bankerRed: String?, bankerBlue: String?): Int {
        val cfg = playConfig(type)
        val reds = splitNums(red)
        val blues = splitNums(blue)
        val bRed = splitNums(bankerRed)
        val bBlue = splitNums(bankerBlue)
        val rc = if (bRed.isNotEmpty()) combinations(diff(reds, bRed), cfg.minRed - bRed.size).size
        else combinations(reds, cfg.minRed).size
        val bc = if (bBlue.isNotEmpty()) combinations(diff(blues, bBlue), cfg.minBlue - bBlue.size).size
        else combinations(blues, cfg.minBlue).size
        return rc * bc
    }

    /**
     * 计算单注奖金描述。一等奖/二等奖为浮动奖，其余为固定奖金。
     * 大乐透在奖池≥8亿时，三至七等奖自动上浮（2026新规）。
     */
    fun prizeMoney(type: String, tier: String, poolAmount: Long = 0): String {
        if (type == "ssq") {
            return when (tier) {
                "一等奖" -> "浮动（最高500万）"
                "二等奖" -> "浮动"
                "三等奖" -> "3000元"
                "四等奖" -> "200元"
                "五等奖" -> "10元"
                "六等奖" -> "5元"
                "福运奖" -> "5元"
                else -> ""
            }
        } else {
            val boost = poolAmount >= 8_0000_0000 // 8亿元
            return when (tier) {
                "一等奖" -> "浮动（最高1000万）"
                "二等奖" -> "浮动"
                "三等奖" -> if (boost) "6666元" else "5000元"
                "四等奖" -> if (boost) "380元" else "300元"
                "五等奖" -> if (boost) "200元" else "150元"
                "六等奖" -> if (boost) "18元" else "15元"
                "七等奖" -> if (boost) "7元" else "5元"
                else -> ""
            }
        }
    }

    /** 把奖池金额（元）格式化为可读字符串，如"6.29亿" */
    fun poolAmountDesc(poolAmount: Long): String {
        if (poolAmount <= 0) return ""
        val yi = poolAmount.toDouble() / 1_0000_0000.0
        return if (yi >= 1) {
            String.format("%.2f亿", yi)
        } else {
            val wan = poolAmount.toDouble() / 1_0000.0
            String.format("%.0f万", wan)
        }
    }

    private val TIER_RANK = mapOf(
        "一等奖" to 1, "二等奖" to 2, "三等奖" to 3, "四等奖" to 4,
        "五等奖" to 5, "六等奖" to 6, "七等奖" to 7, "福运奖" to 8
    )

    data class TicketMatch(
        val won: Boolean, val tier: String, val bestMr: Int, val bestMb: Int,
        val bets: Int, val hitRed: List<String>, val hitBlue: List<String>, val matchText: String,
        val fyjHint: String = ""
    )

    fun matchTicket(type: String, red: String?, blue: String?, bankerRed: String?, bankerBlue: String?, drawRed: String?, drawBlue: String?, poolAmount: Long = 0): TicketMatch {
        val bets = ticketBets(type, red, blue, bankerRed, bankerBlue)
        val combos = enumerateCombos(type, red, blue, bankerRed, bankerBlue)
        var best = ""
        var bestMr = 0
        var bestMb = 0
        for ((cRed, cBlue) in combos) {
            val r = matchTier(type, cRed.joinToString(","), cBlue.joinToString(","), drawRed, drawBlue, poolAmount)
            if (r.mr > bestMr || (r.mr == bestMr && r.mb > bestMb)) {
                bestMr = r.mr
                bestMb = r.mb
            }
            if (r.tier.isNotEmpty() && (best.isEmpty() || (TIER_RANK[r.tier] ?: 99) < (TIER_RANK[best] ?: 99))) {
                best = r.tier
            }
        }
        val hb = hitBalls(red, blue, drawRed, drawBlue)
        // 福运奖提醒：双色球3红+0蓝且无奖池数据时，提示可能符合福运奖
        var fyjHint = ""
        if (type == "ssq" && bestMr == 3 && bestMb == 0 && best.isEmpty() && poolAmount == 0L) {
            fyjHint = "可能符合福运奖条件（需本期奖池≥15亿，请查看福彩官网确认）"
        }
        return TicketMatch(
            won = best.isNotEmpty(), tier = best, bestMr = bestMr, bestMb = bestMb,
            bets = bets, hitRed = hb.hitRed, hitBlue = hb.hitBlue,
            matchText = "命中 ${bestMr}+${bestMb}",
            fyjHint = fyjHint
        )
    }

    /** 等级徽标配色（与网页 TIER_STYLE 对应）。返回 null 表示无对应样式。 */
    data class TierStyle(val bg: Int, val fg: Int)

    fun tierStyle(tier: String): TierStyle? = when (tier) {
        "一等奖" -> TierStyle(0x2Ec8182e.toInt(), 0xffc0152f.toInt())
        "二等奖" -> TierStyle(0x26d8343f.toInt(), 0xffd8343f.toInt())
        "三等奖" -> TierStyle(0x24e85a5f.toInt(), 0xffd8343f.toInt())
        "四等奖" -> TierStyle(0x21f07a7e.toInt(), 0xffd8343f.toInt())
        "五等奖" -> TierStyle(0x1ff59a9d.toInt(), 0xffd8343f.toInt())
        "六等奖" -> TierStyle(0x1cfab8ba.toInt(), 0xffc0152f.toInt())
        "七等奖" -> TierStyle(0x1afac8c8.toInt(), 0xffc0152f.toInt())
        "福运奖" -> TierStyle(0x26ffc107.toInt(), 0xffb8860b.toInt())
        else -> null
    }

    /** 老数据 prize_tier 可能存数字（"5" / "5等奖"），这里按数字映射到等级样式。 */
    fun tierNumStyle(num: String): TierStyle? {
        val n = num.toIntOrNull() ?: return null
        return when (n) {
            1 -> tierStyle("一等奖")
            2 -> tierStyle("二等奖")
            3 -> tierStyle("三等奖")
            4 -> tierStyle("四等奖")
            5 -> tierStyle("五等奖")
            6 -> tierStyle("六等奖")
            7 -> tierStyle("七等奖")
            8 -> tierStyle("福运奖")
            else -> null
        }
    }
}

File: web/src/utils/match.js
// 中奖匹配与等级计算（纯前端，依据现有 /api/lottery 与 /api/draw 数据实时计算）
// 规则与中国福利彩票 / 体彩官方一致。

export function splitNums(s) {
  return (s || '')
    .split(',')
    .map((x) => x.trim())
    .filter(Boolean)
}

export function intersectCount(a, b) {
  const set = new Set(b)
  return a.filter((x) => set.has(x)).length
}

// 命中的号码（用于球上打标）
export function hitBalls(userRed, userBlue, drawRed, drawBlue) {
  const dr = new Set(splitNums(drawRed))
  const db = new Set(splitNums(drawBlue))
  return {
    hitRed: splitNums(userRed).filter((x) => dr.has(x)),
    hitBlue: splitNums(userBlue).filter((x) => db.has(x)),
  }
}

// 计算中奖等级。返回 { mr, mb, tier, won }
// tier: ''（未中奖）/ '一等奖' ... '七等奖' / '福运奖'
// poolAmount: 该期奖池金额（元），用于判断福运奖是否生效（SSQ奖池≥15亿才触发）
export function matchTier(type, userRed, userBlue, drawRed, drawBlue, poolAmount = 0) {
  const mr = intersectCount(splitNums(userRed), splitNums(drawRed))
  const mb = intersectCount(splitNums(userBlue), splitNums(drawBlue))
  let tier = ''
  if (type === 'ssq') {
    // 双色球（2026新规）：福运奖只在奖池≥15亿时生效
    const fyjActive = poolAmount >= 1500000000 // 15亿元
    if (mr === 6 && mb === 1) tier = '一等奖'
    else if (mr === 6 && mb === 0) tier = '二等奖'
    else if (mr === 5 && mb === 1) tier = '三等奖'
    else if ((mr === 5 && mb === 0) || (mr === 4 && mb === 1)) tier = '四等奖'
    else if ((mr === 4 && mb === 0) || (mr === 3 && mb === 1)) tier = '五等奖'
    else if (mb === 1) tier = '六等奖'
    else if (mr === 3 && mb === 0 && fyjActive) tier = '福运奖'
  } else {
    // 大乐透（2026新规7等奖级）
    if (mr === 5 && mb === 2) tier = '一等奖'
    else if (mr === 5 && mb === 1) tier = '二等奖'
    else if ((mr === 5 && mb === 0) || (mr === 4 && mb === 2)) tier = '三等奖'
    else if (mr === 4 && mb === 1) tier = '四等奖'
    else if ((mr === 4 && mb === 0) || (mr === 3 && mb === 2)) tier = '五等奖'
    else if ((mr === 3 && mb === 1) || (mr === 2 && mb === 2)) tier = '六等奖'
    else if ((mr === 3 && mb === 0) || (mr === 2 && mb === 1) || (mr === 1 && mb === 2) || (mr === 0 && mb === 2)) tier = '七等奖'
  }
  return { mr, mb, tier, won: tier !== '' }
}

// ---------- 复式 / 胆拖 展开 ----------

export function playConfig(type) {
  if (type === 'dlt') {
    return { minRed: 5, maxRed: 12, minBlue: 2, maxBlue: 12, redCount: 35, blueCount: 12, name: '大乐透', redLabel: '前区', blueLabel: '后区' }
  }
  return { minRed: 6, maxRed: 20, minBlue: 1, maxBlue: 16, redCount: 33, blueCount: 16, name: '双色球', redLabel: '红球', blueLabel: '蓝球' }
}

function combinations(arr, k) {
  const res = []
  const n = arr.length
  if (k < 0 || k > n) return res
  const idx = new Array(k)
  const rec = (start, i) => {
    if (i === k) {
      res.push(idx.map((j) => arr[j]))
      return
    }
    for (let j = start; j < n; j++) {
      idx[i] = j
      rec(j + 1, i + 1)
    }
  }
  rec(0, 0)
  return res
}

function diff(a, b) {
  const set = new Set(b)
  return a.filter((x) => !set.has(x))
}

// enumerateCombos 把复式/胆拖展开为所有单式组合，返回 [{red:[],blue:[]}]
export function enumerateCombos(type, red, blue, bankerRed, bankerBlue) {
  const cfg = playConfig(type)
  const reds = splitNums(red)
  const blues = splitNums(blue)
  const bRed = splitNums(bankerRed)
  const bBlue = splitNums(bankerBlue)
  let redCombos, blueCombos
  if (bRed.length) {
    redCombos = combinations(diff(reds, bRed), cfg.minRed - bRed.length).map((c) => [...bRed, ...c])
  } else {
    redCombos = combinations(reds, cfg.minRed)
  }
  if (bBlue.length) {
    blueCombos = combinations(diff(blues, bBlue), cfg.minBlue - bBlue.length).map((c) => [...bBlue, ...c])
  } else {
    blueCombos = combinations(blues, cfg.minBlue)
  }
  const out = []
  for (const rc of redCombos) for (const bc of blueCombos) out.push({ red: rc, blue: bc })
  return out
}

// ticketBets 计算注数（不含倍数）
export function ticketBets(type, red, blue, bankerRed, bankerBlue) {
  const cfg = playConfig(type)
  const reds = splitNums(red)
  const blues = splitNums(blue)
  const bRed = splitNums(bankerRed)
  const bBlue = splitNums(bankerBlue)
  let rc = 1
  let bc = 1
  if (bRed.length) rc = combinations(diff(reds, bRed), cfg.minRed - bRed.length).length
  else rc = combinations(reds, cfg.minRed).length
  if (bBlue.length) bc = combinations(diff(blues, bBlue), cfg.minBlue - bBlue.length).length
  else bc = combinations(blues, cfg.minBlue).length
  return rc * bc
}

const TIER_RANK = { 一等奖: 1, 二等奖: 2, 三等奖: 3, 四等奖: 4, 五等奖: 5, 六等奖: 6, 七等奖: 7, 福运奖: 8 }

// matchTicket 处理复式/胆拖，返回最佳中奖等级与命中球（命中球取所选号码与开奖号的交集）
// poolAmount: 该期奖池金额（元），用于判断福运奖是否生效及大乐透奖金升级
export function matchTicket(type, red, blue, bankerRed, bankerBlue, drawRed, drawBlue, poolAmount = 0) {
  const bets = ticketBets(type, red, blue, bankerRed, bankerBlue)
  const combos = enumerateCombos(type, red, blue, bankerRed, bankerBlue)
  let best = ''
  let bestMr = 0
  let bestMb = 0
  for (const c of combos) {
    const r = matchTier(type, c.red.join(','), c.blue.join(','), drawRed, drawBlue, poolAmount)
    // 不管是否中奖，都记录最大命中数（用于 UI 展示"命中 X+Y"）
    if (r.mr > bestMr || (r.mr === bestMr && r.mb > bestMb)) {
      bestMr = r.mr
      bestMb = r.mb
    }
    // 只在中奖时更新最佳等级
    if (r.tier && (!best || TIER_RANK[r.tier] < TIER_RANK[best])) {
      best = r.tier
    }
  }
  // 命中球：所选号码中出现在开奖结果里的（直观展示哪些号中了）
  const dr = new Set(splitNums(drawRed))
  const db = new Set(splitNums(drawBlue))
  const hitRed = splitNums(red).filter((x) => dr.has(x))
  const hitBlue = splitNums(blue).filter((x) => db.has(x))
  return {
    won: best !== '',
    tier: best,
    bestMr,
    bestMb,
    bets,
    hitRed,
    hitBlue,
    matchText: `命中 ${bestMr}+${bestMb}`,
  }
}

// 计算单注奖金描述。一等奖/二等奖为浮动奖，其余为固定奖金。
// 大乐透在奖池≥8亿时，三至七等奖自动上浮（2026新规）。
export function prizeMoney(type, tier, poolAmount = 0) {
  if (type === 'ssq') {
    switch (tier) {
      case '一等奖': return '浮动（最高500万）'
      case '二等奖': return '浮动'
      case '三等奖': return '3000元'
      case '四等奖': return '200元'
      case '五等奖': return '10元'
      case '六等奖': return '5元'
      case '福运奖': return '5元'
      default: return ''
    }
  } else {
    const boost = poolAmount >= 800000000 // 8亿元
    switch (tier) {
      case '一等奖': return '浮动（最高1000万）'
      case '二等奖': return '浮动'
      case '三等奖': return boost ? '6666元' : '5000元'
      case '四等奖': return boost ? '380元' : '300元'
      case '五等奖': return boost ? '200元' : '150元'
      case '六等奖': return boost ? '18元' : '15元'
      case '七等奖': return boost ? '7元' : '5元'
      default: return ''
    }
  }
}

// 把奖池金额（元）格式化为可读字符串，如"6.29亿"
export function poolAmountDesc(poolAmount) {
  if (!poolAmount || poolAmount <= 0) return ''
  const yi = poolAmount / 100000000
  if (yi >= 1) return yi.toFixed(2) + '亿'
  const wan = poolAmount / 10000
  return Math.round(wan) + '万'
}

// 等级徽标配色（统一红色主题，按深浅区分：一等奖最深 → 六等奖最浅）
export const TIER_STYLE = {
  一等奖: { bg: 'rgba(200,24,46,0.18)', fg: '#c0152f', ring: 'rgba(200,24,46,0.45)' },
  二等奖: { bg: 'rgba(216,52,63,0.15)', fg: '#d8343f', ring: 'rgba(216,52,63,0.42)' },
  三等奖: { bg: 'rgba(232,90,95,0.14)', fg: '#d8343f', ring: 'rgba(216,52,63,0.38)' },
  四等奖: { bg: 'rgba(240,122,126,0.13)', fg: '#d8343f', ring: 'rgba(216,52,63,0.34)' },
  五等奖: { bg: 'rgba(245,154,157,0.12)', fg: '#d8343f', ring: 'rgba(216,52,63,0.30)' },
  六等奖: { bg: 'rgba(250,184,186,0.11)', fg: '#c0152f', ring: 'rgba(216,52,63,0.26)' },
  七等奖: { bg: 'rgba(250,200,200,0.10)', fg: '#c0152f', ring: 'rgba(216,52,63,0.22)' },
  福运奖: { bg: 'rgba(255,193,7,0.15)', fg: '#b8860b', ring: 'rgba(255,193,7,0.40)' },
}
